**회귀 분석**
- 사례) 최종 성적(Y)을 출석률(X1)과 과제점수(X2)로 설명



In [1]:
import pandas as pd

df = pd.DataFrame({
    "출석률": [90, 85, 88, 75, 80, 95, 70, 65, 92, 78],
    "과제점수": [85, 80, 82, 70, 75, 90, 68, 60, 88, 72],
    "성적": [88, 82, 85, 74, 78, 92, 70, 65, 90, 76]
})

In [5]:
pip install statsmodels

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 25.6 MB/s eta 0:00:0000:010:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.3/233.3 kB 32.7 MB/s eta 0:00:00

[notice] A new release of pip available: 22.3.1 -> 26.1.2
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


1. 다중회귀분석 수행

In [8]:
import statsmodels.formula.api as smf

# 종속변수: 성적
# 독립변수: 출석률, 과제
model = smf.ols('성적 ~ 출석률 + 과제점수', data=df).fit() #oldirnay least Sqaures(최소제곱법)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                     성적   R-squared:                       0.998
Model:                            OLS   Adj. R-squared:                  0.997
Method:                 Least Squares   F-statistic:                     1502.
Date:                Sat, 20 Jun 2026   Prob (F-statistic):           6.06e-10
Time:                        00:56:18   Log-Likelihood:                -5.2390
No. Observations:                  10   AIC:                             16.48
Df Residuals:                       7   BIC:                             17.39
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0613      1.396      5.058      0.0

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/scipy/stats/_axis_nan_policy.py:430: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=10 observations were given.
  return hypotest_fun_in(*args, **kwds)


# 풀이

In [10]:
print(dir(smf))

['__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'conditional_logit', 'conditional_mnlogit', 'conditional_poisson', 'gee', 'glm', 'glmgam', 'gls', 'glsar', 'logit', 'mixedlm', 'mnlogit', 'negativebinomial', 'nominal_gee', 'ols', 'ordinal_gee', 'phreg', 'poisson', 'probit', 'quantreg', 'rlm', 'wls']


In [9]:
import statsmodels.formula.api as smf

# 종속변수: 성적
# 독립변수: 출석률, 과제점수
model = smf.ols('성적 ~ 출석률 + 과제점수', data=df).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                     성적   R-squared:                       0.998
Model:                            OLS   Adj. R-squared:                  0.997
Method:                 Least Squares   F-statistic:                     1502.
Date:                Sat, 20 Jun 2026   Prob (F-statistic):           6.06e-10
Time:                        00:57:31   Log-Likelihood:                -5.2390
No. Observations:                  10   AIC:                             16.48
Df Residuals:                       7   BIC:                             17.39
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0613      1.396      5.058      0.0

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/scipy/stats/_axis_nan_policy.py:430: UserWarning: `kurtosistest` p-value may be inaccurate with fewer than 20 observations; only n=10 observations were given.
  return hypotest_fun_in(*args, **kwds)


2. 결과해석

(1) 결정계수(R-squared, Adj. R-squared)
- R-squared  = 0.998
  → 전체 성적 변동의 99.8%를 출석률과 과제점수 두 변수로 설명 가능
- Adj. R-squared = 0.997
  → 설명변수 개수를 보정한 값으로도 99.7% 설명 가능

(2) 회귀계수 (coef)
- Intercept = 7.0613
- 출석률 = 0.5512 (p=0.007)
  → 출석률이 1% 증가할 때 성적은 약 0.55점 증가
  → p < 0.05 이므로 통계적으로 유의미한 변수
- 과제점수 = 0.3617 (p=0.046)
  → 과제점수가 1점 증가할 때 성적은 약 0.36점 증가
  → p < 0.05 이므로 통계적으로 유의미한 변수

(3) 결론
- 회귀식 : 성적 = 7.061 + 0.551 x 출석률 + 0.362 x 과제점수


3. 상관계수 구하기

In [ ]:
df[["출석률", "과제점수"]].corr()

,출석률,과제점수
출석률,1.000000,0.993559
과제점수,0.993559,1.000000


In [ ]:
print('스피어만 상관 계수 : \n', df[["출석률", "과제점수"]].corr(method='spearman'))
print('켄달 상관 계수 : \n', df[["출석률", "과제점수"]].corr(method='kendall'))

스피어만 상관 계수 : 
       출석률  과제점수
출석률   1.0   1.0
과제점수  1.0   1.0
켄달 상관 계수 : 
       출석률  과제점수
출석률   1.0   1.0
과제점수  1.0   1.0


4. 출석률 82점, 과제점수 80점에 대해, 구축한 회귀 모델에서의 예상 성적 구하기

In [14]:
new_data = pd.DataFrame({"출석률": [82], "과제점수": [80]})
pred = model.predict(new_data)
print("예상성적", pred[0])

예상성적 81.19539061176081


In [ ]:
new_data = pd.DataFrame({"출석률": [82], "과제점수": [80]})
pred = model.predict(new_data)
print("예상 성적:", pred[0])

예상 성적: 81.19539061176053
